### ISSUE TITLE
Could not install libraries in Linux

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#9583]
### Description
I would like to check frequently new changes on OpenModelica, through nightly builds. I think the best way to do this is to install it in Linux, with nightly updates.

So, I followed the instructions from 
https://openmodelica.org/download/download-linux

even though I have near-zero experience with Linux, and therefore might have made some basic error.

Using those instructions, I  installed OpenModelica on  an Ubuntu 22.04  (in a VirtualBox VM) using:

echo \
 "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/openmodelica-keyring.gpg] https://build.openmodelica.org/apt \
 $(lsb_release -cs) release" | sudo tee /etc/apt/sources.list.d/openmodelica.list > /dev/null

OM got installed, but when I open it, Modelica is not loaded (screenshot 1).
If I try to install it through OpenModelica's library manager, I get an error message (screenshot 2)
If I try to install any other library (e.g. PowerGrids) I get the same error message (screenshot 3)

### Steps to Reproduce
See description

### Expected Behavior
I expected Modelica to be automatically loaded, and that loading additional libraries works smoothly.
But I'm not sure if I made some basic mistake in installing OM, although the error messages do not give me any hint about it.

### Screenshots
Screenshot1:
![image](https://user-images.githubusercontent.com/7373030/197338495-b2d6d94a-0cf4-4ba2-812c-61f08c7df0d8.png)

Screenshot2:
![image](https://user-images.githubusercontent.com/7373030/197338520-ef803b68-4307-40e9-9e52-8e322683b424.png)

Screenshot3:
![image](https://user-images.githubusercontent.com/7373030/197338595-50b18dca-2d45-47e9-9d16-2ba7f18ce824.png)

### Version and OS
See description

**Comment by casella** [comment]
@ceraolo we have fundamentally reworked package management for the upcoming 1.20.0 release, see #8291. The idea is that upon installation, the MSL is automatically installed with the package manager in one's .openmodelica directory as soon as any OM tool tries to load it. This is expected to work out of the box, and also to allow running off-line or behind-firewalls installations.

The new infrastructure is ready and tested for Windows and you can get it with any recent nightly build. It still needs to be packaged in the Linux installers, see #8291. In fact, that is the last critical item we need to get done before releasing 1.20.0.

We should be able to get that done next week, and as soon as that is done you can experiment with the latest nightly or with the 1.20.0-dev.beta1 release. Please monitor #8291 for updates.

**Comment by perost** [comment]
Using the package manager should not fail though, so something is wrong here. Does the `/home/massimo/.openmodelica/cache` folder exist? It seems odd that it would fail to find the file like that.

**Comment by ceraolo** [comment]
not sure how to check. I did as follows

![image](https://user-images.githubusercontent.com/7373030/197348020-ffb931df-ec08-4f08-8be4-cdd5668f3f64.png)

Sorry for being so dumb with Ubuntu

**Comment by ceraolo** [comment]
I also tried  the Ubuntu 22.04 pre-built VM from
https://build.openmodelica.org/omc/builds/virtual_machines/base_image/
but to my surprise I discovered that there is not OpenModelica installed inside.

**Comment by casella** [comment]
@perost it could be that the fact that loadModel doesn't find the cached MSL when looking for it for the first time puts it in some bad state.

I would really suggest that we fix that first.

**Comment by casella** [comment]
> I also tried the Ubuntu 22.04 pre-built VM from https://build.openmodelica.org/omc/builds/virtual_machines/base_image/ but to my surprise I discovered that there is not OpenModelica installed inside.

Huh? That should be fixed for sure. Would you mind opening a ticket for that?

**Comment by ceraolo** [comment]
> > I also tried the Ubuntu 22.04 pre-built VM from https://build.openmodelica.org/omc/builds/virtual_machines/base_image/ but to my surprise I discovered that there is not OpenModelica installed inside.
> 
> Huh? That should be fixed for sure. Would you mind opening a ticket for that?

I'll do

**Comment by ceraolo** [comment]
BTW, yesterday I updated my Ubuntu installation and now OM has Modelica 4.0.0 installed.
Trying to install PowerGrids fails as before, though.

**Comment by perost** [comment]
> @perost it could be that the fact that loadModel doesn't find the cached MSL when looking for it for the first time puts it in some bad state.
> 
> I would really suggest that we fix that first.

That was my guess too, but I tried deleting my .openmodelica folder and install a package without having any cached files and that worked just fine. Which makes sense, the installation of cached packages is only done if it can actually find the cached index which we don't yet distribute in the Linux builds.

So there's something else going on here. I guess I'll have to install Ubuntu myself and see if I can replicate the issue.

**Comment by perost** [comment]
Setting this to blocker for 1.12.0 since we really need to look into this before doing the release.

**Comment by sjoelund** [comment]
> > I also tried the Ubuntu 22.04 pre-built VM from https://build.openmodelica.org/omc/builds/virtual_machines/base_image/ but to my surprise I discovered that there is not OpenModelica installed inside.
> 
> Huh? That should be fixed for sure. Would you mind opening a ticket for that?

There's not supposed to be OMC in the base image (it's the image without OM; just a normal Ubuntu that we use to install OM onto the other images)

**Comment by sjoelund** [comment]
As for the actual issue; could it be that ~/.openmodelica exists and is owned by root or something?
```
chown -R $USER ~/.openmodelica/
```

**Comment by perost** [comment]
I installed Ubuntu 22.04 in VirtualBox and installed OM as per the instructions, and I get the same issue. It actually creates the `~/.openmodelica/cache/74d056ef1b4a89463408ba70fc4ea4db83a6296f.zip` file, but the file is empty.

If I manually download the file and put it in the cache folder then it installs just fine. So it seems like something goes wrong with the download for some reason, but I have no idea what. And considering that @ceraolo says he now has MSL 4.0.0 it seems like it sometimes actually does work, which is even more odd.

**Comment by perost** [comment]
When trying to install PowerGrids it correctly downloads the zip file for PowerGrids, but then it first tries to install MSL 3.2.3 which fails because the zip file is again empty. Installing Buildings 9.0.0 works fine, since that uses MSL 4.0.0 which I as mentioned installed by downloading the zip manually.

So from my limited testing it seems like the issue only affects the MSL zip files for some reason.

**Comment by perost** [comment]
@ceraolo: Also, I noticed in the command you posted that you used `release` instead of `nightly`, so if that's the command you used then you're not actually using the nightly builds. I made sure to use the nightly build myself though, so the issue certainly exists there at least.

**Comment by sjoelund** [comment]
> It actually creates the `~/.openmodelica/cache/74d056ef1b4a89463408ba70fc4ea4db83a6296f.zip` file, but the file is empty.

We probably should download it as `~/.openmodelica/cache/74d056ef1b4a89463408ba70fc4ea4db83a6296f.zip.tmp` and then rename it when it succeeds...

**Comment by ceraolo** [comment]
> > > I also tried the Ubuntu 22.04 pre-built VM from https://build.openmodelica.org/omc/builds/virtual_machines/base_image/ but to my surprise I discovered that there is not OpenModelica installed inside.
> > 
> > 
> > Huh? That should be fixed for sure. Would you mind opening a ticket for that?
> 
> There's not supposed to be OMC in the base image (it's the image without OM; just a normal Ubuntu that we use to install OM onto the other images)

Ok, I expected to find "OMEdit" ready-to use among the available, installed, applications. If there is OMEdit, there is also OMC (and not vice-versa)

Probably it is what the majority of users expects.

As far as I'm concerned this is not an issue, since I prefer installing myself on a pure Ubuntu (instead of xubuntu). So for me, it is ok if you decide to leave things as they are,  or decide to make OMEdit readily available.

**Comment by ceraolo** [comment]
> @ceraolo: Also, I noticed in the command you posted that you used `release` instead of `nightly`, so if that's the command you used then you're not actually using the nightly builds. I made sure to use the nightly build myself though, so the issue certainly exists there at least.

My mistake in the description of the ticket: I'm going to edit it.

BTW, 
In the docu I see in bold the word **stable** in the command, and "Or replace **stable**, **nightly** or **release**" in the instructions. 

I conclude that I must write:

 "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/openmodelica-keyring.gpg] https://build.openmodelica.org/apt \
 $(lsb_release -cs) nightly" | sudo tee /etc/apt/sources.list.d/openmodelica.list > /dev/null

i.e. `lsb_release` should remain as such, isn't it?

**Comment by perost** [comment]
> > > > I also tried the Ubuntu 22.04 pre-built VM from https://build.openmodelica.org/omc/builds/virtual_machines/base_image/ but to my surprise I discovered that there is not OpenModelica installed inside.
> > > 
> > > 
> > > Huh? That should be fixed for sure. Would you mind opening a ticket for that?
> > 
> > 
> > There's not supposed to be OMC in the base image (it's the image without OM; just a normal Ubuntu that we use to install OM onto the other images)
> 
> Ok, I expected to find "OMEdit" ready-to use among the available, installed, applications. If there is OMEdit, there is also OMC (and not vice-versa)
> 
> Probably it is what the majority of users expects.
> 
> As far as I'm concerned this is not an issue, since I prefer installing myself on a pure Ubuntu (instead of xubuntu). So for me, it is ok if you decide to leave things as they are, or decide to make OMEdit readily available.

I think his point is that the `base_image` images aren't meant to be used by users, they are used by us to produce the other images. You should instead use e.g. `https://build.openmodelica.org/omc/builds/virtual_machines/nightly/` if you want a VM with OM on it.

**Comment by perost** [comment]
> i.e. `lsb_release` should remain as such, isn't it?

Yes, `lsb_release -cs` will print out which Ubuntu version you're using, e.g. `jammy` for 22.04 (Jammy Jellyfish).

**Comment by casella** [comment]
> There's not supposed to be OMC in the base image (it's the image without OM; just a normal Ubuntu that we use to install OM onto the other images)

@ceraolo, in fact, if you check https://build.openmodelica.org/omc/builds/virtual_machines/base_image/, the README.txt file reads:
> Contains the base image we use to create the VM images. These images do not contain OpenModelica. For that, check the stable, nightly or release folders.

**Comment by casella** [comment]
@ceraolo/@perost, coming back to the original issue of this ticket, can you please try
```
chown -R $USER ~/.openmodelica/
```
and report what you get?

**Comment by ceraolo** [comment]
> > There's not supposed to be OMC in the base image (it's the image without OM; just a normal Ubuntu that we use to install OM onto the other images)
> 
> @ceraolo, in fact, if you check https://build.openmodelica.org/omc/builds/virtual_machines/base_image/, the README.txt file reads:
> 
> > Contains the base image we use to create the VM images. These images do not contain OpenModelica. For that, check the stable, nightly or release folders.

OK.

**Comment by ceraolo** [comment]
> @ceraolo/@perost, coming back to the original issue of this ticket, can you please try
> 
> ```
> chown -R $USER ~/.openmodelica/
> ```
> 
> and report what you get?

![VirtualBox_ubuntu1_24_10_2022_17_59_59](https://user-images.githubusercontent.com/7373030/197571982-99820b39-adb4-4b73-ac92-fcd650c49475.png)

**Comment by perost** [comment]
> @ceraolo/@perost, coming back to the original issue of this ticket, can you please try
> 
> ```
> chown -R $USER ~/.openmodelica/
> ```
> 
> and report what you get?

That command will just set the owner of `.openmodelica` to the user, so it won't actually show anything. But it's not a permission issue, otherwise files wouldn't be created at all. At least for me the zip-files are created as they should, but nothing is downloaded for the MSL libraries so they end up empty.

**Comment by ceraolo** [comment]
> When trying to install PowerGrids it correctly downloads the zip file for PowerGrids, but then it first tries to install MSL 3.2.3 which fails because the zip file is again empty. Installing Buildings 9.0.0 works fine, since that uses MSL 4.0.0 which I as mentioned installed by downloading the zip manually.

Great analysis! I checked myself and I could install Buildings as well,

**Comment by sjoelund** [comment]
> At least for me the zip-files are created as they should, but nothing is downloaded for the MSL libraries so they end up empty.

Yeah, then it would be best to download as .tmp and remove target + rename the file on success. It might just be github that's very slow?

**Comment by sjoelund** [comment]
If you create a 0-size file it will never try to redownload. So that's a problem. The second issue would be to try to add more error messages to the curl module when it fails

**Comment by perost** [comment]
> If you create a 0-size file it will never try to redownload. So that's a problem. The second issue would be to try to add more error messages to the curl module when it fails

Yeah, that would be an improvement. But I just tested it on Ubuntu 22.04 by trying to install MSL 3.2.3 10 times while deleting the empty zip-file before each try, and it failed every time. Then I tried the same on Arch Linux and it worked the first time.

**Comment by sjoelund** [comment]
For me, it works fine in Ubuntu 22.04

**Comment by perost** [comment]
> For me, it works fine in Ubuntu 22.04

Maybe it's VirtualBox that's the issue then?

**Comment by sjoelund** [comment]
I guess we need more error-messages to find out why. It's a very big difference in time to download for me. 4 minutes, 3 seconds, 16 seconds, etc. Might be an issue with the networking stack for virtualbox or just the added time that is the problem there. I tried on my home network (native Ubuntu 22.04) and from LiU (WSL2 Ubuntu 22.04).

**Comment by sjoelund** [comment]
It's not a big piece of code to debug, but if I cannot reproduce it, I guess I can't do it...
https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/runtime/om_curl.c

**Comment by perost** [comment]
> I guess we need more error-messages to find out why. It's a very big difference in time to download for me. 4 minutes, 3 seconds, 16 seconds, etc. Might be an issue with the networking stack for virtualbox or just the added time that is the problem there. I tried on my home network (native Ubuntu 22.04) and from LiU (WSL2 Ubuntu 22.04).

For me it's consistently 4-5 seconds to download and install MSL 4.0.0 via the package manager (deleting the whole .openmodelica folder between tries).

**Comment by casella** [comment]
In any case, the MSL is arguably the largest (or among the largest) libraries around, which can take longer time to download from github (which I also experienced to be somewhat slower recently).

Allow me to recall our priorities. We should first fix the Linux installation so it does not require to download such large files from the network in the first place, and so that we can finally proceed with the 1.20.0 release. If someone still experiences this kind of problems afterwards when downloading other libraries from a virtual machine, we can address the problem with due time.

**Comment by perost** [comment]
#9597 should hopefully fix the issue.

It turns out that we added the same URL multiple times to the list of files to download, because one zip-file can contain multiple libraries. So we'd try to download e.g. Modelica and its dependencies ModelicaServices  and Complex, and all of them are from the same zip-file.

Making sure that the list of downloads only contains unique entries seems to fix the issue, at least for me. It might be because the downloaded file is closed once the download is complete, but I'm not sure why that would cause the file to be end up empty and only in some cases. Some quirk with the VirtualBox virtual harddrive perhaps? I don't know, but at least it seems to work fine now.

**Comment by perost** [comment]
Just a heads up that the [Linux builds](https://test.openmodelica.org/jenkins/job/LINUX_BUILDS/) seems to broken for the moment, so it might take a while before the fix is available in the nightly builds.
